### ISSUE TITLE
Inconsistent octant usage between training and inference for phi/theta embedding

### ISSUE THREAD

**Issue by nebulajo** [Choddeok/EmoSphere-TTS#12]
Hi, thanks for your great work on EmoSphere-TTS! I’ve been experimenting with the code implementation, and I have a few questions.

During training, it seems phi and theta are embedded only for points in the lower hemisphere (dominance < 0), specifically octants V, VII, VII, VIII, based on the bin definitions:

```python
self.azimuth_bins = nn.Parameter(
    torch.linspace(-np.pi / 2, np.pi, 4), requires_grad=False
)
self.azimuth_emb = Embedding(4, self.hidden_size // 8)

self.elevation_bins = nn.Parameter(
    torch.linspace(np.pi / 2, np.pi, 2), requires_grad=False
)
self.elevation_emb = Embedding(2, self.hidden_size // 8)
```

However, during inference, the code uses octants I, III, V, VII instead:
```python
if hparams['emo_style'] == "I":
    emo_VAD_style[0] = torch.tensor([np.pi/4, np.pi/4])
elif hparams['emo_style'] == "III":
    emo_VAD_style[0] = torch.tensor([np.pi/4, -3*np.pi/4])
elif hparams['emo_style'] == "V":
    emo_VAD_style[0] = torch.tensor([3*np.pi/4, np.pi/4])
elif hparams['emo_style'] == "VII":
    emo_VAD_style[0] = torch.tensor([3*np.pi/4, -3*np.pi/4])
```

Shouldn’t inference also use the same octants as training (V, VII, VII, VIII) for consistency?

Additionally, I’m wondering whether emotion labels (e.g., Happy, Sad) should be used independently from phi/theta values. For example, if I select the Sad label during inference, does it matter which octant (I, III, V, VII) is used? Or should certain emo_styles be preferred for specific emotions?

**Comment by Choddeok** [comment]
torch.bucketize(azimuth, self.azimuth_bins) determines which bin (interval) each azimuth value belongs to based on the bin boundaries defined in self.azimuth_bins, and returns the bin index as an integer.


---
# 1.How it works
azimuth_index = torch.bucketize(azimuth, self.azimuth_bins)

- azimuth: Actual azimuth angle values (e.g., [-1.0, 0.5, 2.0] in radians).

- self.azimuth_bins: The predefined bin boundaries.

torch.linspace(-np.pi/2, np.pi, 4)
# tensor([-1.5708,  0.0000,  1.5708,  3.1416])

- torch.bucketize checks each azimuth value and returns the index of the bin it belongs to.

- The default behavior (right=False) means the left edge of the bin is inclusive.

- Returned indices range from 0 to len(bins).

# 2.Example
import torch, math
bins = torch.linspace(-math.pi/2, math.pi, 4)  
# tensor([-1.5708,  0.0000,  1.5708,  3.1416])

azimuth = torch.tensor([-2.0, -1.0, 0.5, 2.5])
azimuth_index = torch.bucketize(azimuth, bins)
Explanation:

- 2.0 → smaller than bins[0] → index 0

- 1.0 → between -1.5708 and 0 → index 1

- 0.5 → between 0 and 1.5708 → index 2

- 2.5 → between 1.5708 and 3.1416 → index 3
print(azimuth_index)  
# tensor([0, 1, 2, 3])

**Comment by Choddeok** [comment]
Thank you for your question.
As explained, the code actually uses all eight octants, not just the lower hemisphere. Both azimuth (4 bins) and elevation (2 bins) embeddings cover the full set of octants when combined.


torch.bucketize(azimuth, self.azimuth_bins) determines which bin (interval) each azimuth value belongs to based on the bin boundaries defined in self.azimuth_bins, and returns the bin index as an integer.


---
# 1.How it works
```
azimuth_index = torch.bucketize(azimuth, self.azimuth_bins)
```
- azimuth: Actual azimuth angle values (e.g., [-1.0, 0.5, 2.0] in radians).

- self.azimuth_bins: The predefined bin boundaries.
```
torch.linspace(-np.pi/2, np.pi, 4)
# tensor([-1.5708,  0.0000,  1.5708,  3.1416])
```
- torch.bucketize checks each azimuth value and returns the index of the bin it belongs to.

- The default behavior (right=False) means the left edge of the bin is inclusive.

- Returned indices range from 0 to len(bins).

# 2.Example
```
import torch, math
bins = torch.linspace(-math.pi/2, math.pi, 4)  
# tensor([-1.5708,  0.0000,  1.5708,  3.1416])

azimuth = torch.tensor([-2.0, -1.0, 0.5, 2.5])
azimuth_index = torch.bucketize(azimuth, bins)
```
Explanation:

- -2.0 → smaller than bins[0] → index 0

- -1.0 → between -1.5708 and 0 → index 1

- 0.5 → between 0 and 1.5708 → index 2

- 2.5 → between 1.5708 and 3.1416 → index 3

**Comment by nebulajo** [comment]
Thank you so much for the detailed explanation! 🙏 It helped me understand how the binning with torch.bucketize works.
